src/darwin/Framework/CHIP/MTRDeviceConnectivityMonitor.mm

```diff
@@ -192,100 +242,157 @@ static void ResolveCallback(
-        MTR_LOG("%@ connectivity monitor already running", self);
+    // If monitor is nil, object was deallocated - callback safely ignored
+    if (!monitor) {
         return;
     }
 
-    MTR_LOG("%@ start connectivity monitoring for %@ (%lu monitoring objects)", self, _instanceName, static_cast<unsigned long>(sConnectivityMonitorCount));
+    [monitor handleResolvedHostname:hostName port:port error:errorCode];
+}
 
-    auto sharedConnection = [MTRDeviceConnectivityMonitor _sharedResolverConnection];
-    if (!sharedConnection) {
-        MTR_LOG_ERROR("%@ failed to get shared resolver connection", self);
-        return;
-    }
+- (BOOL)startMonitoringWithHandler:(MTRDeviceConnectivityMonitorHandler)handler queue:(dispatch_queue_t)queue
+{
+    // Initialize sSharedResolverQueue on first use
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sSharedResolverQueue = dispatch_queue_create("MTRDeviceConnectivityMonitor", DISPATCH_QUEUE_SERIAL_WITH_AUTORELEASE_POOL);
+    });
+
+    __block BOOL result = NO;
 
-    for (auto domain : kResolveDomains) {
-        DNSServiceRef resolver = sharedConnection;
-        DNSServiceErrorType dnsError = DNSServiceResolve(&resolver,
-            kDNSServiceFlagsShareConnection,
-            kDNSServiceInterfaceIndexAny,
-            _instanceName.UTF8String,
-            kOperationalType,
-            domain,
-            ResolveCallback,
-            (__bridge void *) self);
-        if (dnsError == kDNSServiceErr_NoError) {
-            _resolvers.emplace_back(std::move(resolver));
-        } else {
-            MTR_LOG_ERROR("%@ failed to create resolver for \"%s\" domain: %" PRId32, self, StringOrNullMarker(domain), dnsError);
+    // Run entire function on sSharedResolverQueue to maintain lock ordering (queue → lock)
+    // and avoid deadlocks with callbacks that also use this pattern.
+    dispatch_sync(sSharedResolverQueue, ^{
+        std::lock_guard lock(sConnectivityMonitorLock);
+
+        _monitorHandler = handler;
+        _handlerQueue = queue;
+
+        MTR_LOG("%@ start connectivity monitoring for %@ (current monitoring objects: %lu)", self, _instanceName, static_cast<unsigned long>(sConnectivityMonitorCount));
+
+        // If there's already a resolver running, just return
+        if (_resolvers.size() != 0) {
+            MTR_LOG("%@ connectivity monitor already running", self);
+            result = YES;
+            return;
         }
-    }
 
-    if (_resolvers.size() != 0) {
-        sConnectivityMonitorCount++;
-    }
+        auto sharedConnection = [MTRDeviceConnectivityMonitor _sharedResolverConnection];
+        if (!sharedConnection) {
+            MTR_LOG_ERROR("%@ failed to get shared resolver connection", self);
+            _monitorHandler = nil;
+            _handlerQueue = nil;
+            result = NO;
+            return;
+        }
+
+        // We're already on the connection's dispatch queue, so call DNSServiceResolve directly
```

Comment:
Ah, is this why we need to be on the queue?  Because calling the DNS APIs needs to happen on the queue?  That wasn't obvious to me from the DNS API documentation.....

Reply:
As I said above - it's not obvious, but when I didn't have this, I occasionally (1 in 20 times) ran into broken pipe errors during repeated testing.